### ISSUE TITLE
[storage-transfer] Can not use custom service account

### ISSUE THREAD

**Issue by ivansenic** [googleapis/google-cloud-java#9247]
Setting the `com.google.api.gax.core.CredentialsProvider` in the `StorageTransferServiceSettings` has no effect. Althougt I confirmed that credentials are correctly initialized, the `storage-transfer` client ignores this. 

Setup:

```
CredentialsProvider credentialsProvider;
StorageTransferServiceSettings.Builder builder = StorageTransferServiceSettings.newBuilder().setCredentialsProvider(credentialsProvider);
StorageTransferServiceSettings transferServiceSettings = builder.build();

try (StorageTransferServiceClient transferClient = StorageTransferServiceClient.create(transferServiceSettings)) {
   // define job
   ....
    transferClient.createTransferJob(request);
}
```

Credentials: 

```
2023-03-17 18:03:29.614  INFO 2529 --- [           main] c.g.c.s.a.c.GcpContextAutoConfiguration  : The default project ID is my-project-hidden
2023-03-17 18:03:29.723  INFO 2529 --- [           main] c.g.c.s.core.DefaultCredentialsProvider  : Default credentials provider for service account hidden@hidden.iam.gserviceaccount.com
2023-03-17 18:03:29.724  INFO 2529 --- [           main] c.g.c.s.core.DefaultCredentialsProvider  : Scopes in use by default credentials: [https://www.googleapis.com/auth/pubsub, https://www.googleapis.com/auth/spanner.admin, https://www.googleapis.com/auth/spanner.data, https://www.googleapis.com/auth/datastore, https://www.googleapis.com/auth/sqlservice.admin, https://www.googleapis.com/auth/devstorage.read_only, https://www.googleapis.com/auth/devstorage.read_write, https://www.googleapis.com/auth/cloudruntimeconfig, https://www.googleapis.com/auth/trace.append, https://www.googleapis.com/auth/cloud-platform, https://www.googleapis.com/auth/cloud-vision, https://www.googleapis.com/auth/bigquery, https://www.googleapis.com/auth/monitoring.write]
```

Result:

```
Caused by: io.grpc.StatusRuntimeException: FAILED_PRECONDITION: Failed to obtain the location of the GCS bucket my-bucket Additional details: project-myprojectid@storage-transfer-service.iam.gserviceaccount.com does not have storage.buckets.get access to the Google Cloud Storage bucket. Permission 'storage.buckets.get' denied on resource (or it may not exist).

```

Using the same `CredentialsProvider` for the Storage works without issues. The service account used has all needed permissions.

#### Environment details

1. Specify the API at the beginning of the title. For example, "[vision]: ...").
   General, Core, and Other are also allowed as types
2. OS type and version: Linux
3. Java version: java 17
4. Version(s): latest

#### Steps to reproduce

See above.

#### Code example

See above.

#### Stack trace
```
com.google.api.gax.rpc.FailedPreconditionException: io.grpc.StatusRuntimeException: FAILED_PRECONDITION: Failed to obtain the location of the GCS bucket plainly-user-projects-dev Additional details: project-518619120753@storage-transfer-service.iam.gserviceaccount.com does not have storage.buckets.get access to the Google Cloud Storage bucket. Permission 'storage.buckets.get' denied on resource (or it may not exist).
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:102)
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:41)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:86)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:67)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1132)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1270)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1038)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:808)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:574)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:544)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:541)
	at io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:489)
	at io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:453)
	at io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:486)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:576)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:757)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:736)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.storagetransfer.v1.proto.StorageTransferServiceClient.createTransferJob(StorageTransferServiceClient.java:314)
```

#### External references such as API reference guides
None.

#### Any additional information below
The worst thing is that the service account given in the error message is not available in the GCP IAM at all. It exists, I confirmed it here: https://cloud.google.com/storage-transfer/docs/reference/rest/v1/googleServiceAccounts/get, but there is no way the needed permissions can be added.

So basically the Java library is useless at the moment. I don't know how anybody can use this.

**Comment by emmileaf** [comment]
@ivansenic Thanks for reporting this! I believe `project-myprojectid@storage-transfer-service.iam.gserviceaccount.com` is the [google-managed service account](https://cloud.google.com/iam/docs/service-agents) that is picked up from Google Application Default Credentials, so something is likely missing with the `CredentialsProvider` override. Could you help share some more information to help us narrow down and reproduce this issue? In particular:

1. How are you passing in the service account credentials as `CredentialsProvider`? (e.g. through `com.google.api.gax.core.FixedCredentialsProvider`, `com.google.cloud.spring.core.DefaultCredentialsProvider`, etc?)
2. Are you importing and interacting with the `java-storage-transfer` library directly, or through the [spring-cloud-gcp autoconfigurations](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/tree/main/spring-cloud-previews)? (similar question for `java-storage`, since you mentioned it working as expected)
3. On your `StorageTransferServiceClient` object, could you try calling/printing `.getSettings().getCredentialsProvider().getCredentials()`, and seeing whether that returns the expected service account credentials? Similarly, for the working `Storage` object, does calling `.getOptions().getCredentials()` return the same credentials or something different?

**Comment by ivansenic** [comment]
@emmileaf Sure, let me try to provide help and answer to all questions.

## 1. Dependencies

>     1. How are you passing in the service account credentials as `CredentialsProvider`? (e.g. through `com.google.api.gax.core.FixedCredentialsProvider`, `com.google.cloud.spring.core.DefaultCredentialsProvider`, etc?)

I am using Spring starter approach. I have following in my `build.gradle` (including the bom management in version `3.4.6` :

```
// gcp
implementation 'com.google.cloud:spring-cloud-gcp-starter-logging'
implementation 'com.google.cloud:spring-cloud-gcp-starter-storage'
implementation 'com.google.cloud:google-cloud-storage-transfer'
```

Later on I have property `spring.cloud.gcp.credentials.encoded-key=..` set. This auto-configures the  the `CredentialsProvider credentialsProvider`, that I can autowire and use like:

```
    @Bean
    @ConditionalOnProperty("spring.cloud.gcp.core.enabled")
    public GcpStorageService gcpStorageService(@Value("${spring.cloud.gcp.project-id}") String projectId, Storage storage, CredentialsProvider credentialsProvider) throws IOException {
        // transfer
        StorageTransferServiceSettings.Builder builder = StorageTransferServiceSettings.newBuilder().setCredentialsProvider(credentialsProvider);
        StorageTransferServiceSettings transferServiceSettings = builder.build();

        return new GcpStorageService(projectId, storage, transferServiceSettings);
    }
```

The `GcpStorageService` is my own implementation, where I use the `transferServiceSettings` to create the transfer client:

```
try (StorageTransferServiceClient transferClient = StorageTransferServiceClient.create(transferServiceSettings)) {
   // define job
   ....
    transferClient.createTransferJob(request);
}
```

Here one important thing to note is that for some reason I need to specify the project id in the transfer job. Without this the API call fails, although I guess the project id is anyway defined in the service account I use, so this is a bit suspicious.

```
TransferTypes.TransferJob job = TransferTypes.TransferJob.newBuilder()
                    .setProjectId(gcpProjectId)
                    .setStatus(TransferTypes.TransferJob.Status.ENABLED)
                    .setTransferSpec(spec)
                    .build();
```

## 2. Strater

>     2. Are you importing and interacting with the `java-storage-transfer` library directly, or through the [spring-cloud-gcp autoconfigurations](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/tree/main/spring-cloud-previews)? (similar question for `java-storage`, since you mentioned it working as expected)

I was importing `java-storage-transfer` directly, because I was not aware that the starter exists. This could be the reason.

The storage works both ways. I was able to use the starter just fine, but before I had a starter I used it directly using smth like:

```
ByteArrayInputStream privateFileStream = new ByteArrayInputStream(privateFileKey.getBytes());
  GoogleCredentials credentials = GoogleCredentials.fromStream(privateFileStream).createScoped(Collections.singletonList("https://www.googleapis.com/auth/cloud-platform"));
  Storage service = StorageOptions.newBuilder().setCredentials(credentials).build().getService();
```

## 3. Debug

>     3. On your `StorageTransferServiceClient` object, could you try calling/printing `.getSettings().getCredentialsProvider().getCredentials()`, and seeing whether that returns the expected service account credentials? Similarly, for the working `Storage` object, does calling `.getOptions().getCredentials()` return the same credentials or something different?

The both methods return the same object (same identity), with correct service account set.

![Screenshot from 2023-03-21 12-16-55](https://user-images.githubusercontent.com/10600041/226590846-b6a9d4e0-0c3c-493d-ac1b-30122c750c80.png)

**Comment by emmileaf** [comment]
@ivansenic Thanks for the additional info! Don’t worry about the switch between using java-storage-transfer directly vs. the starter, that was more of a question to help narrow down sources of possible error. 

Since the credentials are being correctly set on the `StorageTransferServiceClient` object, I suspect the issue could be from project configuration rather than the Java client library, and that the project might be missing permissions specific to using the Storage Transfer API. Could you take a look at this [product documentation on transfer permissions](https://cloud.google.com/storage-transfer/docs/iam-cloud) and if it provides any help to your scenario?

In particular, you can check for missing permissions in your project by running `gcloud transfer authorize`, seeing if it includes an entry like below, and follow the product documentation to update any missing permissions.
```
Google-managed transfer account project-id@storage-transfer-service.iam.gserviceaccount.com has roles:
['roles/storage.admin', 'roles/storagetransfer.serviceAgent']
Missing roles:
[]
```

For further help on this issue, please reach out to [Google Cloud Support](https://cloud.google.com/support). Thanks!
